bin/evm2near/src/wasm_translate.rs

```diff
@@ -0,0 +1,749 @@
+                    DataSegmentMode::Active {
+                        memory_index,
+                        offset: const_exprs.last().unwrap(),
+                    }
+                }
+                DataMode::Passive => DataSegmentMode::Passive,
+            };
+            let data_segment = DataSegment { mode, data: d.data };
+            data_section.segment(data_segment);
+        }
+        m.section(&data_section);
+
+        m
+    }
+}
+
+pub fn parse(wasm: &Vec<u8>) -> Result<ModuleBuilder> {
+    let parsed = wasmparser::Parser::new(0)
+        .parse_all(wasm.as_slice())
+        .map(|p| p.unwrap())
+        .collect::<Vec<_>>();
+
+    let mut code_section_size: Option<u32> = None;
+
+    let mut builder = ModuleBuilder::new();
+    for p in parsed {
+        match p {
+            Payload::Version {
+                num: _version,
+                encoding,
+                range: _range,
+            } => {
+                assert_eq!(encoding, wasmparser::Encoding::Module);
+            }
+            Payload::TypeSection(type_section) => {
+                for typ in type_section {
+                    let (params, results) = type_def(typ?)?;
+                    builder.types.push(Signature { params, results });
+                }
+            }
+            Payload::ImportSection(import_section) => {
+                for import in import_section {
+                    let import = import?;
+                    let typ = type_ref(import.ty)?;
+                    builder.imports.push(Import {
+                        module: import.module.to_string(),
+                        field: import.name.to_string(),
+                        ty: typ,
+                    });
+                }
+            }
+            Payload::FunctionSection(function_section) => {
+                for function in function_section {
+                    builder.functions.push(function?);
+                }
+            }
+            Payload::TableSection(table_section) => {
+                for table in table_section {
+                    let table = table?;
+                    if let wasmparser::TableInit::RefNull = table.init {
+                    } else {
+                        panic!("unknown table init");
+                    }
+                    let table_type = table_type(&table.ty)?;
+                    builder.tables.push(table_type);
+                }
+            }
+            Payload::MemorySection(memory_section) => {
+                for mem in memory_section {
+                    let mem = mem?;
+                    let memory_type = memory_type(&mem)?;
+                    builder.memories.push(memory_type);
+                }
+            }
+            Payload::TagSection(tag_section) => {
+                for tag in tag_section {
+                    let tag = tag?;
+                    let _tag_type = tag_type(&tag)?;
+                    panic!("unspecified section (spec doesnt have that one! WTF)");
```

Comment:
I wonder if we should make an issue to come back to this when the spec is updated?

Reply:
as far as i understand, its near wasm runtime limitation at the moment
wasm-core-2 actually support [data count section](https://www.w3.org/TR/wasm-core-2/binary/modules.html#data-count-section)
and tag section appears to be supported in [exception proposal](https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md)
so probably we should follow near runtime instead of wasm spec itself